controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -259,54 +268,95 @@ public void refresh() {
     }
 
     @Override
-    CompletableFuture<Void> createNewTransaction(final UUID txId, final long timestamp, final long leaseExpiryTime,
-                                                 final long maxExecutionExpiryTime,
-                                                 final long scaleGracePeriod) {
-        final String activePath = getActiveTxPath(txId.toString());
-        return store.createZNodeIfNotExist(activePath,
-                new ActiveTxnRecord(timestamp, leaseExpiryTime, maxExecutionExpiryTime, scaleGracePeriod, TxnStatus.OPEN)
-                        .toByteArray())
-                .thenApply(x -> cache.invalidateCache(activePath));
-    }
-
-    @Override
-    CompletableFuture<Data<Integer>> getActiveTx(final UUID txId) {
-        final String activeTxPath = getActiveTxPath(txId.toString());
-
+    CompletableFuture<Integer> createNewTransaction(final UUID txId,
+                                                    final long timestamp,
+                                                    final long leaseExpiryTime,
+                                                    final long maxExecutionExpiryTime,
+                                                    final long scaleGracePeriod) {
+        CompletableFuture<Integer> future = new CompletableFuture<>();
+        createNewTransactionNode(txId, timestamp, leaseExpiryTime, maxExecutionExpiryTime, scaleGracePeriod)
+                .whenComplete((value, ex) -> {
+                    if (ex != null) {
+                        if (ExceptionHelpers.getRealException(ex) instanceof DataNotFoundException) {
+                            FutureHelpers.completeAfter(() -> createNewTransactionNode(txId, timestamp, leaseExpiryTime,
+                                    maxExecutionExpiryTime, scaleGracePeriod), future);
+                        } else {
+                            future.completeExceptionally(ex);
+                        }
+                    } else {
+                        future.complete(value);
+                    }
+                });
+        return future;
+    }
+
+    private CompletableFuture<Integer> createNewTransactionNode(final UUID txId,
+                                                                final long timestamp,
+                                                                final long leaseExpiryTime,
+                                                                final long maxExecutionExpiryTime,
+                                                                final long scaleGracePeriod) {
+        return getLatestEpoch().thenCompose(pair -> {
+            final String activePath = getActiveTxPath(pair.getKey(), txId.toString());
+            final byte[] txnRecord = new ActiveTxnRecord(timestamp, leaseExpiryTime, maxExecutionExpiryTime,
+                    scaleGracePeriod, TxnStatus.OPEN).toByteArray();
+            // TODO: replace the previous create version with `createZNodeIfParentExists` method, once scale operation
```

Comment:
please do the failure handling when parent node doesnt exist. So that get active epoch is called again and this is retried on new epoch by calling getLatestEpoch.